lib/unison-sqlite/src/Unison/Sqlite/Transaction.hs

```diff
@@ -0,0 +1,225 @@
+module Unison.Sqlite.Transaction
+  ( -- * Transaction management
+    Transaction,
+    runTransaction,
+
+    -- * Executing queries
+
+    -- ** Without results
+
+    -- *** With parameters
+    execute,
+    executeMany,
+
+    -- *** Without parameters
+    execute_,
+
+    -- ** With results
+
+    -- *** With parameters
+    queryList,
+    queryListOne,
+    queryMaybe,
+    queryMaybeOne,
+    queryOne,
+    queryOneOne,
+
+    -- **** With checks
+    queryListCheck,
+    queryListOneCheck,
+    queryMaybeCheck,
+    queryMaybeOneCheck,
+    queryOneCheck,
+    queryOneOneCheck,
+
+    -- *** Without parameters
+    queryList_,
+    queryListOne_,
+    queryMaybe_,
+    queryMaybeOne_,
+    queryOne_,
+    queryOneOne_,
+
+    -- **** With checks
+    queryListCheck_,
+    queryListOneCheck_,
+    queryMaybeCheck_,
+    queryMaybeOneCheck_,
+    queryOneCheck_,
+    queryOneOneCheck_,
+  )
+where
+
+import Control.Monad.Trans.Reader (ReaderT (..))
+import qualified Database.SQLite.Simple as Sqlite
+import qualified Database.SQLite.Simple.FromField as Sqlite
+import Unison.Prelude
+import Unison.Sqlite.Connection (Connection (..))
+import qualified Unison.Sqlite.Connection as Connection
+import Unison.Sqlite.Sql
+
+newtype Transaction a
+  = Transaction (Connection -> IO a)
+  -- Omit MonadIO instance because transactions may be retried
+  -- Omit MonadThrow instance so we always throw SqliteException (via *Check) with lots of context
+  deriving (Applicative, Functor, Monad) via (ReaderT Connection IO)
+
+-- | Run a transaction on the given connection.
+runTransaction :: MonadIO m => Connection -> Transaction a -> m a
+runTransaction conn@(Connection _ _ conn0) (Transaction f) =
+  -- TODO some sensible retry logic
+  liftIO do
+    Sqlite.execute_ conn0 "BEGIN"
+    result <- f conn
+    Sqlite.execute_ conn0 "COMMIT"
```

Comment:
> We should probably be bracketing all of our transactions, even if the interface of Transaction doesn't allow failures I think we still need to handle async exceptions and the only safe way is with some form of bracketing.

Since this transaction type isn't even used yet, I punted on the implementation here to keep the diff smaller. But I will follow up with retry + async exception safety logic in another PR.

> Also, I'm wondering if we should just always use savepoints, since those have better nesting properties. Otherwise we need to always worry whether we're currently nested within an outer transaction, which is hard to reason about.

Hmm... I don't understand why we would need to worry whether we're currently nested in an outer transaction. The type of `runTransaction` does not allow it:

```haskell
runTransaction :: MonadIO m => Connection -> Transaction a -> m a
```

> IMO, a withTransaction helper similar to the one from sqlite-simple is the safest way to avoid foot-guns. We can lift the IO using unliftIO as necessary.

I don't understand, sorry. Safe in what way? What are the foot-guns?

Reply:
> Since this transaction type isn't even used yet, I punted on the implementation here to keep the diff smaller.

I don't think I understand, this is all new code isn't it? I don't see how rewriting this to be exception safe changes anything to do with existing code :thinking:

To be clear, I'm thinking it should look something like this:

```
  liftIO do
    bracketOnError 
      (Sqlite.execute_ conn0 "BEGIN")
      (\_ -> f conn0 <* Sqlite.execute_ conn0 "COMMIT")
      (\_ -> Sqlite.execute_ conn0 "ROLLBACK")
```

Reply:
> Am I missing something that handles that correctly elsewhere?

It's true, there's no exception safety code here. I intentionally left `runTransaction` as a stub just to keep the diff smaller. (Note: also not included here is actually using this library; this PR just adds it to the monorepo but doesn't port the existing code over yet).